xbmc/addons/InputStream.cpp

```diff
@@ -0,0 +1,506 @@
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "InputStream.h"
+#include "utils/StringUtils.h"
+#include "utils/log.h"
+#include "cores/VideoPlayer/DVDDemuxers/DVDDemux.h"
+
+
+namespace ADDON
+{
+
+std::unique_ptr<CInputStream> CInputStream::FromExtension(AddonProps props, const cp_extension_t* ext)
+{
+  std::string listitemprops = CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@listitemprops");
+  std::string extensions = CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@extension");
+  std::string name(ext->plugin->identifier);
+  return std::unique_ptr<CInputStream>(new CInputStream(std::move(props),
+                                                        std::move(name),
+                                                        std::move(listitemprops),
+                                                        std::move(extensions)));
+}
+
+CInputStream::CInputStream(AddonProps props, std::string name, std::string listitemprops, std::string extensions)
+: InputStreamDll(std::move(props))
+{
+  m_fileItemProps = StringUtils::Tokenize(listitemprops, "|");
+  for (auto &key : m_fileItemProps)
+  {
+    StringUtils::Trim(key);
+    key = name + "." + key;
+  }
+
+  m_extensionsList = StringUtils::Tokenize(extensions, "|");
+  for (auto &ext : m_extensionsList)
+  {
+    StringUtils::Trim(ext);
+  }
+}
+
+AddonPtr CInputStream::Clone() const
+{
+  // Copy constructor is generated by compiler and calls parent copy constructor
+  return AddonPtr(new CInputStream(*this));
+}
+
+bool CInputStream::Supports(CFileItem &fileitem)
+{
+  std::string extension = URIUtils::GetExtension(fileitem.GetPath());
+  bool match = false;
+  for (auto &ext : m_extensionsList)
+  {
+    if (ext == extension)
+    {
+      match = true;
+      break;
+    }
+  }
+  if (!match)
+    return false;
+
+  std::string pathList;
+  try
+  {
+    pathList = m_pStruct->GetPathList();
+  }
+  catch (std::exception &e)
+  {
+    return false;
+  }
+
+  m_pathList = StringUtils::Tokenize(pathList, "|");
+  for (auto &path : m_pathList)
+  {
+    StringUtils::Trim(path);
+  }
+
+  match = false;
+  for (auto &path : m_pathList)
+  {
+    if (path.empty())
+      continue;
+
+    if (fileitem.GetPath().compare(0, path.length(), path) == 0)
+    {
+      match = true;
+      break;
+    }
+  }
+  if (!match)
+    return false;
+
+  return true;
+}
+
+bool CInputStream::Open(CFileItem &fileitem)
+{
+  INPUTSTREAM props;
+  props.m_nCountInfoValues = 0;
+  for (auto &key : m_fileItemProps)
+  {
+    if (fileitem.GetProperty(key).isNull())
+      continue;
+    props.m_ListItemProperties[props.m_nCountInfoValues].m_strKey = key.c_str();
+    props.m_ListItemProperties[props.m_nCountInfoValues].m_strValue = fileitem.GetProperty(key).asString().c_str();
```

Comment:
asString() returns temporary value and can not be passed.
Possible solution:
std::vector<cstring> values;
....
values.push_back(fileitem.GetProperty(key).asString());
props.m_ListItemProperties[props.m_nCountInfoValues].m_strValue = values.back().c_str();